pkg/apis/machine/v1beta1/machineset_cp_webhook.go

```diff
@@ -0,0 +1,84 @@
+package v1beta1
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+
+	admissionv1beta1 "k8s.io/api/admission/v1beta1"
+	"k8s.io/klog"
+	"sigs.k8s.io/controller-runtime/pkg/webhook/admission"
+)
+
+// MachineSetCPHandler validates ControlPlane MachineSet API resources.
+// implements type Handler interface.
+// https://godoc.org/github.com/kubernetes-sigs/controller-runtime/pkg/webhook/admission#Handler
+type MachineSetCPHandler struct {
+	decoder *admission.Decoder
+}
+
+// NewMachineSetCPValidator returns a new MachineSetCPHandler.
+func NewMachineSetCPValidator() (*MachineSetCPHandler, error) {
+	return createMachineSetCPValidator(), nil
+}
+
+func createMachineSetCPValidator() *MachineSetCPHandler {
+	return &MachineSetCPHandler{}
+}
+
+// InjectDecoder injects the decoder.
+func (v *MachineSetCPHandler) InjectDecoder(d *admission.Decoder) error {
+	v.decoder = d
+	return nil
+}
+
+// Handle handles HTTP requests for admission webhook servers.
+func (v *MachineSetCPHandler) Handle(ctx context.Context, req admission.Request) admission.Response {
+	oldMS := &MachineSet{}
+
+	// Delete requests, the req.Object is empty.
+	if err := v.decoder.DecodeRaw(req.OldObject, oldMS); err != nil {
+		return admission.Errored(http.StatusBadRequest, err)
+	}
+
+	klog.V(3).Infof("Validate webhook called for CP MachineSets: %s", oldMS.GetName())
+
+	newMS := &MachineSet{}
+	if req.Operation != admissionv1beta1.Delete {
+		if err := v.decoder.Decode(req, newMS); err != nil {
+			return admission.Errored(http.StatusBadRequest, err)
+		}
+	}
+
+	// Succeed if deleting non-CP MachineSet or Updating a non-CP MachineSet
+	// and the user is not attempting to change it to a CP MachineSet.
+	if !isCPMS(oldMS) && (newMS == nil || !isCPMS(newMS)) {
+		return admission.Allowed("MachineSet is Not Control Plane.")
+	}
+
+	// If the user is updating a CP MachineSet, as long as machine role is
+	// unchanged, we're ok.
+	if newMS != nil && isCPMS(newMS) && isCPMS(oldMS) {
+		return admission.Allowed("Control Plane MachineSet is Valid.")
+	}
+
+	// User is peforming an unallowed operation
+
+	// TODO(michaelgugino): Ensure we account for MachineDeployment ownership
+	// of a CP machineset in the future if we use them.
+	return admission.Denied(fmt.Sprintf("Requested %v of Control Plane MachineSet Not Allowed.", req.Operation))
```

Comment:
Would be nice to have a better error on this, as in, more descriptive of the actual reason why we are rejecting it, I wonder if we could break the logic up so we have a the update and delete handled separately?